addons/product/tests/test_variants.py

```diff
@@ -909,19 +909,103 @@ def test_name_search_dynamic_attributes(self):
+        ProductAttribute = self.env['product.attribute']
+        ProductAttributeValue = self.env['product.attribute.value']
+
+        # Creating attributes
+        dpa_color = ProductAttribute.create({'sequence': 1, 'name': 'color', 'create_variant': 'dynamic'})
+        color_values = ProductAttributeValue.create([{
+            'name': n,
+            'sequence': i,
+            'attribute_id': dpa_color.id,
+        } for i, n in enumerate(['white', 'black'])])
+        dpav_color_white = color_values[0]
+        dpav_color_black = color_values[1]
+
+        dpa_size = ProductAttribute.create({'sequence': 2, 'name': 'size', 'create_variant': 'dynamic'})
+        size_values = ProductAttributeValue.create([{
+            'name': n,
+            'sequence': i,
+            'attribute_id': dpa_size.id,
+        } for i, n in enumerate(['s', 'm'])])
+        dpav_size_s = size_values[0]
+        dpav_size_m = size_values[1]
+
+        # Define a template with only color attribute & white value
+        template = self.env['product.template'].create({
+            'name': 'test product',
+            'attribute_line_ids': [(0, 0, {
+                'attribute_id': dpa_color.id,
+                'value_ids': [(6, 0, [dpav_color_white.id])],
+            })],
+        })
+
+        # Patch unlink method to force archiving instead deleting
+        def unlink(self):
+            self.active = False
+        Product._patch_method('unlink', unlink)
+
+        # Create a variant (because of dynamic attribute)
+        ptav_white = self.env['product.template.attribute.value'].search([
+            ('attribute_line_id', '=', template.attribute_line_ids.id),
+            ('product_attribute_value_id', '=', dpav_color_white.id)
+        ])
+        product_white = Product.with_context(create_product_product=True).create({
+            'product_tmpl_id': template.id,
+            'product_template_attribute_value_ids': [(6, 0, [ptav_white.id])]
+        })
+
+        # Adding a new value to an existing attribute should not archive the variant
+        template.write({
+            'attribute_line_ids': [(1, template.attribute_line_ids[0].id, {
+                'attribute_id': dpa_color.id,
+                'value_ids': [(4, dpav_color_black.id, False)],
+            })]
+        })
+
+        self.assertTrue(product_white.active)
+
+        # Removing an attribute value should archive the product using it
+        template.write({
+            'attribute_line_ids': [(1, template.attribute_line_ids[0].id, {
+                'value_ids': [(3, dpav_color_white.id, 0)],
+            })]
+        })
+
+        self.assertFalse(product_white.active)
+
+        # Reset archiving for the next assert
+        template.write({
+            'attribute_line_ids': [(1, template.attribute_line_ids[0].id, {
+                'value_ids': [(4, dpav_color_white.id, 0)],
+            })]
+        })
+        product_white.active = True
```

Comment:
This is not ok to do manually. If you remove this, and add an assertion that the variant is active again, you basically have your test for the case I mentioned where some archived variants need to be brought back.